process-migration-service/src/main/java/org/kie/processmigration/service/impl/EjbTimerSchedulerServiceImpl.java

```diff
@@ -0,0 +1,79 @@
+package org.kie.processmigration.service.impl;
+
+import java.util.Date;
+import java.util.Optional;
+
+import javax.annotation.PreDestroy;
+import javax.annotation.Resource;
+import javax.ejb.Singleton;
+import javax.ejb.Startup;
+import javax.ejb.Timeout;
+import javax.ejb.Timer;
+import javax.ejb.TimerService;
+import javax.inject.Inject;
+
+import org.kie.processmigration.model.Credentials;
+import org.kie.processmigration.model.Migration;
+import org.kie.processmigration.service.CredentialsService;
+import org.kie.processmigration.service.MigrationService;
+import org.kie.processmigration.service.SchedulerService;
+import org.kie.processmigration.util.DateUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Singleton
+@Startup
+public class EjbTimerSchedulerServiceImpl implements SchedulerService {
+
+    private static final Logger logger = LoggerFactory.getLogger(EjbTimerSchedulerServiceImpl.class);
+
+    @Resource
+    private TimerService timerService;
+
+    @Inject
+    private MigrationService migrationService;
+
+    @Inject
+    private CredentialsService credentialsService;
+
+    @Timeout
+    public void doMigration(Timer timer) {
+        Long migrationId = (Long) timer.getInfo();
+        Migration migration = migrationService.get(migrationId);
+        migrationService.migrate(migration);
+    }
+
+    @Override
+    public void scheduleMigration(Migration migration, Credentials credentials) {
+        Long migrationId = migration.getId();
+        credentialsService.save(credentials.setMigrationId(migrationId));
+        if (migration.getDefinition().getExecution().getScheduledStartTime() == null) {
+            timerService.createTimer(new Date(), migrationId);
+        } else {
+            Date startTime = DateUtils.toDate(migration
+                                                       .getDefinition()
+                                                       .getExecution()
+                                                       .getScheduledStartTime());
+            timerService.createTimer(startTime, migrationId);
+        }
+    }
+
+    @Override
+    public void reScheduleMigration(Migration migration, Credentials credentials) {
+        Optional<Timer> timer = timerService.getTimers()
+                                            .stream()
+                                            .filter(t -> t.getInfo().equals(migration.getId())).findFirst();
+        if (timer.isPresent()) {
+            timer.get().cancel();
+            scheduleMigration(migration, credentials);
+        }
+    }
+
+    @PreDestroy
+    public void stopAll() {
+        for (Timer timer : timerService.getTimers()) {
+            logger.info("Stopping timer: {}", timer.getInfo());
+            timer.cancel();
```

Comment:
wound't this essentially cancel all scheduled migrations? Meaning if server goes down all scheduled migrations are cancelled and thus after server starts again nothing will be scheduled... or am I missing something?

Reply:
No, your assumption is correct. I have removed it.